### PR TITLE
#17736 Composite images construction

### LINKED ISSUE #17736: Restrict data edit setting breaks data grid
### Description

https://user-images.githubusercontent.com/43252954/192403215-10f4d322-fdcf-428f-9fdd-2aa0bf5cd519.mp4

```
Can't switch presentation

Reason:
Argument not valid
```

```
java.lang.IllegalArgumentException: Argument not valid
	at org.eclipse.swt.SWT.error(SWT.java:4899)
	at org.eclipse.swt.SWT.error(SWT.java:4833)
	at org.eclipse.swt.SWT.error(SWT.java:4804)
	at org.eclipse.swt.graphics.ImageData.<init>(ImageData.java:429)
	at org.eclipse.swt.graphics.ImageData.scaledTo(ImageData.java:1139)
	at org.eclipse.swt.internal.DPIUtil.autoScaleImageData(DPIUtil.java:253)
	at org.eclipse.swt.internal.DPIUtil.autoScaleImageData(DPIUtil.java:221)
	at org.eclipse.swt.graphics.Image.getImageData(Image.java:1362)
	at org.eclipse.jface.resource.ImageDataImageDescriptor.getImageData(ImageDataImageDescriptor.java:91)
	at org.eclipse.jface.resource.CompositeImageDescriptor$CachedDescriptorImageDataProvider.getImageData(CompositeImageDescriptor.java:133)
	at org.eclipse.jface.resource.CompositeImageDescriptor$CachedImageDataProvider.computeInPoints(CompositeImageDescriptor.java:89)
	at org.eclipse.jface.resource.CompositeImageDescriptor$CachedImageDataProvider.getHeight(CompositeImageDescriptor.java:73)
	at org.jkiss.dbeaver.ui.OverlayImageDescriptor.<init>(OverlayImageDescriptor.java:35)
	at org.jkiss.dbeaver.ui.DBeaverIcons.getCompositeIcon(DBeaverIcons.java:159)
	at org.jkiss.dbeaver.ui.DBeaverIcons.getIconDescriptor(DBeaverIcons.java:120)
	at org.jkiss.dbeaver.ui.DBeaverIcons.getImage(DBeaverIcons.java:92)
	at org.jkiss.dbeaver.ui.controls.resultset.spreadsheet.SpreadsheetPresentation$GridLabelProvider.getImage(SpreadsheetPresentation.java:2562)
	at org.jkiss.dbeaver.ui.controls.lightgrid.GridColumnRenderer.getColumnImage(GridColumnRenderer.java:67)
	at org.jkiss.dbeaver.ui.controls.lightgrid.GridColumnRenderer.paint(GridColumnRenderer.java:110)
	at org.jkiss.dbeaver.ui.controls.lightgrid.LightGrid.paintColumnsHeader(LightGrid.java:2627)
	at org.jkiss.dbeaver.ui.controls.lightgrid.LightGrid.paintColumnHeader(LightGrid.java:2614)
	at org.jkiss.dbeaver.ui.controls.lightgrid.LightGrid.paintHeader(LightGrid.java:2583)
	at org.jkiss.dbeaver.ui.controls.lightgrid.LightGrid.onPaint(LightGrid.java:2316)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:234)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4251)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1066)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1090)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1075)
	at org.eclipse.swt.widgets.Composite.WM_PAINT(Composite.java:1536)
	at org.eclipse.swt.widgets.Control.windowProc(Control.java:4803)
	at org.eclipse.swt.widgets.Canvas.windowProc(Canvas.java:340)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:5034)
	at org.eclipse.swt.internal.win32.OS.DispatchMessage(Native Method)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3643)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:644)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:551)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:156)
	at org.jkiss.dbeaver.ui.app.standalone.DBeaverApplication.start(DBeaverApplication.java:284)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:596)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1467)
```

### DBeaver Version

22.2.1

### Operating System

win10

### Database and driver

_No response_

### Steps to reproduce

1. Select 'Edit Connection' in connection's context menu
2. Create Connection type with checked setting 'Restrict data edit'
image
3. Select the created type for the connection(reproduces on SQLite, Cocroach; Isn't reproduced on Postgres, Oracle, Clickhouse... )
4. Connect to the database and open data editor

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by E1izabeth**

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/DBeaverIcons.java [hunk 1]
```diff
@@ -173,6 +175,26 @@ private static IconDescriptor getCompositeIcon(IconDescriptor mainIcon, DBIconCo
...
     }
 
+    @NotNull
+    private static ImageDescriptor[] accumulateDecorations(
+        @NotNull DBIconComposite image,
+        @NotNull Function<DBIconComposite, DBPImage> map
+    ) {
+        DBPImage base = image.getMain();
+        if (base instanceof DBIconComposite) {
+            ArrayList<ImageDescriptor> decorations = new ArrayList<>();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/DBeaverIcons.java]:
```suggestion
            List<ImageDescriptor> decorations = new ArrayList<>();
```

File: plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/DBeaverIcons.java [hunk 2]
```diff
@@ -173,6 +175,26 @@ private static IconDescriptor getCompositeIcon(IconDescriptor mainIcon, DBIconCo
...
+            decorations.add(getImageDescriptor(map.apply(image)));
+            do {
+                image = (DBIconComposite) base;
+                decorations.add(getImageDescriptor(map.apply(image)));
+                base = image.getMain();
+            } while (base instanceof DBIconComposite);
+            return decorations.toArray(new ImageDescriptor[decorations.size()]);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui/src/org/jkiss/dbeaver/ui/DBeaverIcons.java]:
```suggestion
            return decorations.toArray(ImageDescriptor[]::new);
```

File: plugins/org.jkiss.dbeaver.tasks.ui.view/plugin.xml [hunk 3]
```diff
@@ -60,7 +60,9 @@
             <with variable="selection">
                 <count value="1"/>
                 <iterate operator="and">
-                    <test property="org.jkiss.dbeaver.task.taskEditable" value="true"/>
+                    <adapt type="org.jkiss.dbeaver.model.task.DBTTask">
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.tasks.ui.view/plugin.xml]: Please remove task-related changes

**Review by ShadelessFox — APPROVED** [review]

**Comment by uslss** [comment]
verified

**Pull request merged**